A2.md

```diff
@@ -0,0 +1,125 @@
+
+This document proposes a mechanism for encoding gRPC service config data
+in DNS for use in the open-source world.
+
+## Background
+
+The [service
+config](https://github.com/grpc/grpc/blob/master/doc/service_config.md)
+mechanism was originally designed for use inside of Google.  However,
+all but one part of the original design will work fine in the open-source
+world.  That one part is the specification of how the service config
+data will be encoded in DNS.
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+There are two parts to this proposal.  The first part is to add some
+JSON wrapping for controlling how service config changes are canary
+tested.  The second part describes how the service config is encoded in
+DNS.
+
+### Canarying Changes
+
+When deploying a change to a service config, it is useful to be able to
+canary test changes to avoid wide-spread breakage by slowly increasing the
+number of clients that see the new version.  To that end, multiple
+service configs choices can be listed, in order, along with criteria that
+determine which choice will be selected by a given client:
+
+```
+// A list of one or more service config choices.
+// The first matching entry wins.
+[
+  {
+    // Criteria used to select this choice.
+    // If a field is absent or empty, it matches all clients.
+    // All fields must match a client for this choice to be selected.
+    //
+    // Client language(s): a list of strings (e.g., 'c++', 'java', 'go',
+    // 'python', etc).
+    'clientLanguage': [string],
+    // Percentage: integer from 0 to 100 indicating the percentage of
+    // clients that should use this choice.
+    'percentage': number,
+    // Client hostname(s): a list of strings.
+    'clientHostname': [string],
+
+    // The service config data object for clients that match the above
+    // criteria.  (The format for this object is defined in
+    // https://github.com/grpc/grpc/blob/master/doc/service_config.md.)
+    'serviceConfig': object
+  }
+]
+```
+
+### Encoding in DNS TXT Records
+
+In DNS, the service config data (in the form documented in the previous
+section) will be encoded in a TXT record via the mechanism described in
+[RFC-1464](https://tools.ietf.org/html/rfc1464) using the attribute name
+`grpc_config`.  The attribute value will be a JSON list containing service
+config choices.  For example, here is an example TXT record:
+
+```
+myserver  3600  TXT  "grpc_config=[{'serviceConfig':{'loadBalancingPolicy':'round_robin','methodConfig':[{'name':[{'service':'MyService','method':'Foo'}],'waitForReady':true}]}}]"
+```
+
+Note that TXT records are limited to 255 bytes per string, as per
+[RFC-1035 section 3.3](https://tools.ietf.org/html/rfc1035#section-3.3).
+However, there can be multiple strings, which will be
+concatenated together, as described in [RFC-4408 section
+3.1.3](https://tools.ietf.org/html/rfc4408#section-3.1.3).  The total
+across all strings is limited to 65536 bytes.  (See the "Open Issues"
```

Comment:
nit: The max size of the entire response is 65535 (not 65536). And in addition to overheads in the header there is also the byte-long length-prefix of each string (maxed out at 256 bytes).

I liked how at the end you mentioned the limit, but in a fuzzy way. Could this sentence be made more fuzzy?